src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1225,73 +1274,138 @@ int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
         ssize_t sent = sendfile(outFd, inFd, NULL, (size >= SSIZE_MAX ? SSIZE_MAX : (size_t)size));
         if (sent < 0)
         {
-            if (errno != EINVAL && errno != ENOSYS)
-            {
-                return -1;
-            }
-            else
+            if (errno == ENOSYS)
             {
-                break;
+                // If sendfile is unavailable in this kernel, size will
+                // remain untouched, so it's fine for concurrent threads to
+                // write to sendFileUnavailable.
+                sendFileUnavailable = true;
             }
+
+            break;
         }
         else
         {
             assert((size_t)sent <= size);
             size -= (size_t)sent;
         }
     }
-    if (size == 0)
+
+    return size;
+}
+#endif // HAVE_SENDFILE_4
+
+#if defined(PAL_COPY_FILE_RANGE_SYSCALL)
+static loff_t CopyFileRange(int inFd, loff_t *inOff, int outFd, loff_t *outOff, size_t len, unsigned int flags)
```

Comment:
Would it make sense to name this `copy_file_range` and define it only if it's not already defined?